stan/math/rev/mat/fun/cholesky_decompose.hpp

```diff
@@ -223,6 +258,135 @@ class cholesky_scalar : public vari {
     }
   }
 };
+#ifdef STAN_OPENCL
+class cholesky_opencl : public vari {
+ public:
+  int M_;
+  vari** variRefA_;
+  vari** variRefL_;
+
+  /**
+   * Constructor for GPU cholesky function.
+   *
+   * Stores varis for A.  Instantiates and stores varis for L.
+   * Instantiates and stores dummy vari for upper triangular part of var
+   * result returned in cholesky_decompose function call
+   *
+   * variRefL aren't on the chainable autodiff stack, only used for storage
+   * and computation. Note that varis for L are constructed externally in
+   * cholesky_decompose.
+   *
+   *
+   * @param A matrix
+   * @param L_A matrix, cholesky factor of A
+   */
+  cholesky_opencl(const Eigen::Matrix<var, -1, -1>& A,
+                  const Eigen::Matrix<double, -1, -1>& L_A)
+      : vari(0.0),
+        M_(A.rows()),
+        variRefA_(ChainableStack::instance().memalloc_.alloc_array<vari*>(
+            A.rows() * (A.rows() + 1) / 2)),
+        variRefL_(ChainableStack::instance().memalloc_.alloc_array<vari*>(
+            A.rows() * (A.rows() + 1) / 2)) {
+    size_t pos = 0;
+    for (size_type j = 0; j < M_; ++j) {
+      for (size_type i = j; i < M_; ++i) {
+        variRefA_[pos] = A.coeffRef(i, j).vi_;
+        variRefL_[pos] = new vari(L_A.coeffRef(i, j), false);
+        ++pos;
+      }
+    }
+  }
+  /**
+   * Reverse mode differentiation algorithm using a GPU
+   *
+   * Reference:
+   *
+   * Iain Murray: Differentiation of the Cholesky decomposition, 2016.
+   *
+   */
+  virtual void chain() {
+    using Eigen::MatrixXd;
+    MatrixXd Lbar_(M_, M_);
+    MatrixXd L_(M_, M_);
+    Lbar_.setZero();
+    L_.setZero();
+
+    size_t pos = 0;
+    for (size_type j = 0; j < M_; ++j) {
+      for (size_type i = j; i < M_; ++i) {
+        Lbar_.coeffRef(i, j) = variRefL_[pos]->adj_;
+        L_.coeffRef(i, j) = variRefL_[pos]->val_;
+        ++pos;
+      }
+    }
+
+    matrix_gpu L(L_);
+    matrix_gpu Lbar(Lbar_);
+    int block_size_
+        = M_ / opencl_context.tuning_opts().cholesky_rev_block_partition;
+    block_size_ = std::max(block_size_, 8);
+    block_size_ = std::min(
+        block_size_, opencl_context.tuning_opts().cholesky_rev_min_block_size);
+    // The following is a GPU implementation of
+    // the chain() function from the cholesky_block
+    // vari class implementation
+    for (int k = M_; k > 0; k -= block_size_) {
+      const int j = std::max(0, k - block_size_);
+      const int k_j_ind = k - j;
+      const int m_k_ind = M_ - k;
+
+      matrix_gpu R(k_j_ind, j);
+      matrix_gpu D(k_j_ind, k_j_ind);
+      matrix_gpu B(m_k_ind, j);
+      matrix_gpu C(m_k_ind, k_j_ind);
+
+      matrix_gpu Rbar(k_j_ind, j);
```

Comment:
So the "bar"s here just refer to sensitivities, an alternative name for which (that we use) is "adjoints," often abbreviated `adj`.

![image](https://user-images.githubusercontent.com/805356/53450439-2f49b180-39ea-11e9-806c-e32c3a4e5be0.png)

Reply:
I forgot to go back and put some more context on this comment. We've talked about trying to use the names from the papers we're implementing as a way to easily go back and forth, but I think we didn't want to talk about that we also want to balance that with the Math library being internally consistent, such that once you understand one area of it you could more easily translate your knowledge into another area. We do this with little things like naming conventions and trying to use the same vocabulary everywhere. But we do want to create a balance - I think in this case a reasonable compromise might be something like `R_adj` here while mentioning the adj -> bar mapping in the doxygen comment containing the reference to the paper. I think that lets a reader go back and forth with the paper while being a little more fluent based on their previous knowledge of the math library. It's definitely a negotiable judgement call but just wanted to explain the two goals we want to balance here.

Reply:
We were more trying to make a 1-to-1 translation of the existing Cholesky_block code than trying to figure out our own names. Thats where both the bar an camel case variRefA come from. I do agree with you that we want to have things go by the naming convention throughout the code base. Can we also change the cholesky_block in this PR or do that separately?

Reply:
Nevermind my question, I have to change it for all vari classes, otherwise `baseVari->vari_ref_L_` wont compile.